src/main/java/io/gingersnapproject/service/MembershipListener.java

```diff
@@ -0,0 +1,131 @@
+package io.gingersnapproject.service;
+
+import java.net.InetAddress;
+import java.net.URI;
+import java.net.UnknownHostException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+import io.gingersnapproject.cdc.ReschedulingTask;
+import io.gingersnapproject.cdc.configuration.Configuration;
+
+import io.quarkus.runtime.ShutdownEvent;
+import io.quarkus.runtime.StartupEvent;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ApplicationScoped
+public class MembershipListener {
+
+   private static final Logger log = LoggerFactory.getLogger(MembershipListener.class);
+   private static final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(r -> new Thread(r,"membership-listener"));
+
+   // Accessed by a single thread.
+   private final Set<InetAddress> members = new HashSet<>();
+   @Inject Configuration configuration;
+
+   // TODO: remove instance wrapper once implementation available.
+   @Inject Instance<DynamicMembershipManagement> dmm;
```

Comment:
Do we expect possibly multiple implementations to listen to membership changes? Should this not just be part of the NotificationManager with a membership change event or something?

Reply:
Besides the `ManagedEngine`, I see only something for metrics. I'll make an event for join and leave.